auto_run.py

```diff
@@ -0,0 +1,146 @@
+
+#import HeatMap
+#imp.reload(HeatMap)
+
+import numpy as np
+
+#%% find duplicates in list
+
+def find_dup(pair_list, element):
+    for (a,b) in pair_list:
+        if element == a or element == b:
+            return True
+    return False
+
+#%% 
+
+def find_epochs(base_dir, sub_dir, epoch_dic, visited_models):
+    directory = os.listdir(os.path.join(base_dir, sub_dir))
+    weight_file = [f for f in directory if (f.endswith(".hdf5"))]
+    for name in weight_file:
+        match = re.search('hpe_epoch(\d+)', name)
+        if match:
+            epoch_dic.append((match.group(1), sub_dir))
+
+    resume_file = [f for f in os.listdir(base_dir) if (sub_dir in f and f is not sub_dir)]
+    for file in resume_file:
+        visited_models.append(file)
+        for name in os.listdir(os.path.join(base_dir, file)):
+            match = re.search('hpe_epoch(\d+)', name)
+            if match and not find_dup(epoch_dic, match.group(1)):
+                epoch_dic.append((match.group(1), file))
+
+    return list(set(epoch_dic))
+
+
+#%% save stacked evaluation heatmaps
+
+def stacked_eval_heatmaps(test_df,eval):
+
+    generator = data_generator.DataGenerator(
+            df=test_df,
+            base_dir=DEFAULT_VAL_IMG_PATH,
+            input_dim=INPUT_DIM,
+            output_dim=OUTPUT_DIM,
+            num_hg_blocks=eval.num_hg_blocks,
+            shuffle=False,  
+            batch_size=len(test_df),
+            online_fetch=False)
+
+    # Select image to predict heatmaps
+    X_batch, y_stacked = generator[0] # There is only one batch in the generator
+    # X_batch, y_stacked = evaluation.load_and_preprocess_img('data/skier.jpg', eval.num_hg_blocks)
+    y_batch = y_stacked[0] # take first hourglass section
+    # Save stacked heatmap images to disk
+    m_batch = test_df.to_dict('records') # TODO: eventually this will be passed from data generator as metadata
+    print("\n\nEval start:   {}\n".format(time.ctime()))
+    eval.visualize_batch(X_batch, y_batch, m_batch)
+    print("\n\nEval end:   {}\n".format(time.ctime()))
+
+#%% 
+
+def predict_kp_bbox(val_df, eval):
+    generator = data_generator.DataGenerator(
+                df=val_df,
+                base_dir=DEFAULT_VAL_IMG_PATH,
+                input_dim=INPUT_DIM,
+                output_dim=OUTPUT_DIM,
+                num_hg_blocks=DEFAULT_NUM_HG,
+                shuffle=False,  
+                batch_size=1,
+                online_fetch=False)
+
+    # Select image to predict heatmaps
+    X_batch, y_stacked = generator[168] # choose one image for evaluation
+    y_batch = y_stacked[0] # take first hourglass section
+    X, y = X_batch[0], y_batch[0] # take first example of batch
+
+    # Get predicted heatmaps for image
+    predict_heatmaps=eval.predict_heatmaps(X_batch)
+
+    # Get predicted keypoints from last hourglass (eval.num_hg_blocks-1)
+    keypoints = eval.heatmaps_to_keypoints(predict_heatmaps[eval.num_hg_blocks-1, 0, :, :, :])
+    print(keypoints)
```

Comment:
Do we want to print the keypoints here?

Reply:
Emmm, not necessary when we are running multiple models but it's very useful when we want to dig into one image. The key point np.array tells us which joint the model is seeing.

Reply:
This is not a callback function. We are running this at the very end. Enabling a callback function would significantly increase our running time.